pkg/network/protocols/tls/types.go

```diff
@@ -0,0 +1,124 @@
+package tls
+
+import (
+	"crypto/tls"
+	"fmt"
+)
+
+// Bitmask constants for Offered_versions matching kernelspace definitions
+const (
+	OfferedTLSVersion10 uint8 = 0x01
+	OfferedTLSVersion11 uint8 = 0x02
+	OfferedTLSVersion12 uint8 = 0x04
+	OfferedTLSVersion13 uint8 = 0x08
+)
+
+// mapping of version constants to their string representations
+var tlsVersionNames = map[uint16]string{
+	tls.VersionTLS10: "tls_1.0",
+	tls.VersionTLS11: "tls_1.1",
+	tls.VersionTLS12: "tls_1.2",
+	tls.VersionTLS13: "tls_1.3",
+}
+
+// Mapping of offered version bitmasks to version constants
+var offeredVersionBitmask = []struct {
+	bitMask uint8
+	version uint16
+}{
+	{OfferedTLSVersion10, tls.VersionTLS10},
+	{OfferedTLSVersion11, tls.VersionTLS11},
+	{OfferedTLSVersion12, tls.VersionTLS12},
+	{OfferedTLSVersion13, tls.VersionTLS13},
+}
+
+// Constants for tag keys
+const (
+	TagTLSVersion       = "tls.version:"
+	TagTLSCipherSuiteID = "tls.cipher_suite_id:"
+	TagTLSClientVersion = "tls.client_version:"
+)
+
+// Tags holds the TLS tags. It is used to store the TLS version, cipher suite and offered versions.
+// We can't use the struct from eBPF as the definition is shared with windows.
+type Tags struct {
+	ChosenVersion   uint16
+	CipherSuite     uint16
+	OfferedVersions uint8
+}
+
+// MergeWith merges the tags from another Tags struct into this one
+func (t *Tags) MergeWith(that Tags) {
+	if t.ChosenVersion == 0 {
+		t.ChosenVersion = that.ChosenVersion
+	}
+	if t.CipherSuite == 0 {
+		t.CipherSuite = that.CipherSuite
+	}
+	if t.OfferedVersions == 0 {
+		t.OfferedVersions = that.OfferedVersions
+	}
+
+}
+
+// IsEmpty returns true if all fields are zero
+func (t *Tags) IsEmpty() bool {
+	return t.ChosenVersion == 0 && t.CipherSuite == 0 && t.OfferedVersions == 0
+}
+
+// String returns a string representation of the Tags struct
+func (t *Tags) String() string {
+	return fmt.Sprintf("ChosenVersion: %d, CipherSuite: %d, OfferedVersions: %d", t.ChosenVersion, t.CipherSuite, t.OfferedVersions)
+}
+
+// FormatTLSVersion converts a version uint16 to its string representation
+func FormatTLSVersion(version uint16) string {
+	if name, ok := tlsVersionNames[version]; ok {
+		return name
+	}
+	return ""
+}
+
+// parseOfferedVersions parses the Offered_versions bitmask into a slice of version strings
+func parseOfferedVersions(offeredVersions uint8) []string {
+	versions := []string{}
+	for _, ov := range offeredVersionBitmask {
+		if (offeredVersions & ov.bitMask) != 0 {
+			if name := tlsVersionNames[ov.version]; name != "" {
+				versions = append(versions, name)
+			}
+		}
+	}
+	return versions
+}
+
+// GetTLSDynamicTags generates dynamic tags based on TLS information
+func GetTLSDynamicTags(tls *Tags) map[string]struct{} {
+	tags := make(map[string]struct{})
+	if tls == nil {
+		return tags
+	}
+
+	// Server chosen version
+	if versionName := FormatTLSVersion(tls.ChosenVersion); versionName != "" {
+		tags[TagTLSVersion+versionName] = struct{}{}
```

Comment:
Since version names are fixed, maybe re-using the tags for them would cut down on allocations here. Ditto for client offered versions.

Reply:
yeah this is a good idea